drivers/serial/uart_ns16550.c

```diff
@@ -484,8 +468,33 @@ static int uart_ns16550_config_get(struct device *dev, struct uart_config *cfg)
  */
 static int uart_ns16550_init(struct device *dev)
 {
+	struct uart_ns16550_dev_data_t * const dev_data = DEV_DATA(dev);
+	const struct uart_ns16550_device_config * const dev_cfg = DEV_CFG(dev);
+
 	int ret;
 
+	ARG_UNUSED(dev_data);
+	ARG_UNUSED(dev_cfg);
+
+#ifdef UART_NS16550_PCIE_ENABLED
+	if (dev_cfg->pcie) {
+		if (!pcie_probe(dev_cfg->pcie_bdf, dev_cfg->pcie_id)) {
+			ret = -EINVAL;
+		}
+
+		dev_data->devconf.port = pcie_get_mbar(dev_cfg->pcie_bdf, 0);
+		pcie_set_cmd(dev_cfg->pcie_bdf, PCIE_CONF_CMDSTAT_MEM, true);
+	}
+#endif
+
+/*
+ * Set port to DEV_DATA, but not overwrite for PCIE enabled
+ * set by pcie_get_mbar() via _configure() above
+ */
+#ifndef UART_NS16550_PCIE_ENABLED
+	DEV_DATA(dev)->devconf.port = DEV_CFG(dev)->devconf.port;
+#endif
```

Comment:
This might be better so it can accommodate platforms with mixed PCIE/non-PCIE UARTs:
```
        } else
#endif
        {
                DEV_DATA(dev)->devconf.port = DEV_CFG(dev)->devconf.port;
        {
```

Reply:
Much better, thanks!